taichi/ui/backends/vulkan/canvas.cpp

```diff
@@ -0,0 +1,276 @@
+      printf("warning, there are no light sources in the scene.\n");
+    }
+    float aspect_ratio =
+        app_context_->swap_chain.swap_chain_extent.width /
+        (float)app_context_->swap_chain.swap_chain_extent.height;
+    scene->update_ubo(aspect_ratio);
+    for (int i = 0; i < scene->mesh_infos_.size(); ++i) {
+      mesh(scene->mesh_infos_[i], scene);
+    }
+    for (int i = 0; i < scene->particles_infos_.size(); ++i) {
+      particles(scene->particles_infos_[i], scene);
+    }
+    scene->mesh_infos_.clear();
+    scene->particles_infos_.clear();
+    scene->point_lights_.clear();
+  } else {
+    throw std::runtime_error("Scene is not vulkan scene");
+  }
+}
+
+void Canvas::cleanup() {
+  for (auto &renderable : renderables_) {
+    renderable->cleanup();
+  }
+  vkDestroySemaphore(app_context_->device(), prev_draw_finished_vk_, nullptr);
+  vkDestroySemaphore(app_context_->device(), this_draw_data_ready_vk_, nullptr);
+}
+
+void Canvas::cleanup_swap_chain() {
+  clear_command_buffer_cache();
+  for (auto &renderable : renderables_) {
+    renderable->cleanup_swap_chain();
+  }
+}
+
+void Canvas::recreate_swap_chain() {
+  for (auto &renderable : renderables_) {
+    renderable->recreate_swap_chain();
+  }
+}
+
+void Canvas::import_semaphores() {
+  if (app_context_->config.ti_arch == Arch::cuda) {
+    prev_draw_finished_cuda_ = (uint64_t)cuda_vk_import_semaphore(
+        prev_draw_finished_vk_, app_context_->device());
+    this_draw_data_ready_cuda_ = (uint64_t)cuda_vk_import_semaphore(
+        this_draw_data_ready_vk_, app_context_->device());
+
+    cuda_vk_semaphore_signal((CUexternalSemaphore)prev_draw_finished_cuda_);
+  }
+}
+
+void Canvas::prepare_for_next_frame() {
+  next_renderable_ = 0;
+  if (app_context_->config.ti_arch == Arch::cuda) {
+    cuda_vk_semaphore_wait((CUexternalSemaphore)prev_draw_finished_cuda_);
+  }
+}
+
+void Canvas::draw_frame(Gui *gui) {
+  uint32_t image_index = app_context_->swap_chain.curr_image_index;
+
+  if (app_context_->swap_chain.images_in_flight[image_index] !=
+      VK_NULL_HANDLE) {
+    vkWaitForFences(app_context_->device(), 1,
+                    &app_context_->swap_chain.images_in_flight[image_index],
+                    VK_TRUE, UINT64_MAX);
+  }
+  app_context_->swap_chain.images_in_flight[image_index] =
+      app_context_->swap_chain
+          .in_flight_scenes[app_context_->swap_chain.current_frame];
+
+  if (app_context_->config.ti_arch == Arch::cuda) {
+    cuda_vk_semaphore_signal((CUexternalSemaphore)this_draw_data_ready_cuda_);
+  }
+
+  VkCommandBuffer command_buffer;
+
+  if (!gui->is_empty) {
+    clear_command_buffer_cache();
+  }
+
+  if (cached_command_buffers_[image_index] != VK_NULL_HANDLE) {
+    command_buffer = cached_command_buffers_[image_index];
+  } else {
+    command_buffer = create_new_command_buffer(app_context_->command_pool(),
+                                               app_context_->device());
```

Comment:
We should build a reuse based transient allocator for these. I have an example is here: https://github.com/bobcao3/BerkeleyGfx/blob/main/sample/2_terrain/terrain.cpp

Reply:
I'll take a look.

Reply:
Is this because we wish to avoid constantly allocating and releasing memories for cmd buffers? If we're to create an allocator for cmd buffers, should we use the same allocator for taichi's vulkan backend as well?